File: src/mirror.py
from typing import ClassVar, Mapping, Sequence, Any, Dict, Optional, Tuple, Final, List, cast
from typing_extensions import Self
from typing import Final

from viam.resource.types import RESOURCE_NAMESPACE_RDK, RESOURCE_TYPE_SERVICE

from viam.module.types import Reconfigurable
from viam.proto.app.robot import ComponentConfig
from viam.proto.common import ResourceName, Vector3
from viam.resource.base import ResourceBase
from viam.resource.types import Model, ModelFamily
from viam.app.viam_client import ViamClient
from viam.rpc.dial import DialOptions
from viam.services.generic import Generic
from viam.logging import getLogger
from viam.utils import ValueTypes, struct_to_dict
from viam.proto.app.data import Filter, TagsFilter
from viam.proto.app.data import BinaryID

import time
import asyncio
from pathlib import Path
import os
import mimetypes
import traceback

LOGGER = getLogger(__name__)

class mirror(Generic, Reconfigurable):
    
    MODEL: ClassVar[Model] = Model(ModelFamily("mcvella", "data"), "mirror")
    
    viam_api_key: str
    viam_api_key_id: str
    labels: list = []
    tags: list =  []
    dataset_id: str = ""
    mirror_path: str = str(Path.home()) + '/.viam/data_mirror'
    app_client: ViamClient
    sync_frequency: int = 60
    running = False
    delete = False

    # Constructor
    @classmethod
    def new(cls, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]) -> Self:
        my_class = cls(config.name)
        my_class.reconfigure(config, dependencies)
        return my_class

    # Validates JSON Configuration
    @classmethod
    def validate(cls, config: ComponentConfig):
        api_key = config.attributes.fields["app_api_key"].string_value
        if api_key == "":
            raise Exception("app_api_key attribute is required")
        api_key_id = config.attributes.fields["app_api_key_id"].string_value
        if api_key_id == "":
            raise Exception("app_api_key_id attribute is required")
        return

    # Handles attribute reconfiguration
    def reconfigure(self, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]):
        self.running = False
        self.dataset_id = config.attributes.fields["dataset_id"].string_value or ""
        self.tags = config.attributes.fields["tags"].list_value or []
        self.labels = config.attributes.fields["labels"].list_value or []
        self.api_key = config.attributes.fields["app_api_key"].string_value or ''
        self.api_key_id = config.attributes.fields["app_api_key_id"].string_value or ''
        self.delete = config.attributes.fields["delete"].bool_value or False
        self.sync_frequency = config.attributes.fields["sync_frequency"].number_value or self.sync_frequency
        mirror_path = config.attributes.fields["mirror_path"].string_value or ''
        if mirror_path != "":
            self.mirror_path =   os.path.join(str(Path.home()) + '/.viam/', mirror_path)
       
        asyncio.ensure_future(self.sync_loop())
        return

    async def sync_loop(self):
        self.app_client: ViamClient = await self.viam_connect()
        self.running = True

        while self.running:
            try:
                await self.do_sync()
                await asyncio.sleep(self.sync_frequency)
            except Exception as e:
                LOGGER.error(f'Error in sync loop: {e}')
                LOGGER.error(traceback.print_exc())
                await asyncio.sleep(1)

    async def do_sync(self):
        # first, get all files and paths current on machine so we 
        # can see if there are extra files that need to be removed
        current_files = {}
        for path, dirs, files in os.walk(self.mirror_path):
            for file in files:
                current_files[os.path.join(self.mirror_path, path, file)] = True
        
        filter_args = {}
        if self.dataset_id != "":
            filter_args['dataset_id'] = self.dataset_id
        if len(self.tags) > 0:
            filter_args['tags_filter'] =  TagsFilter(tags=self.tags)
        filter = Filter(**filter_args)
        if len(self.labels) > 0:
            filter_args['bbox_labels'] = self.labels
        filter = Filter(**filter_args)

        binary_args = {'filter': filter, 'include_binary_data': False}
        
        # we need to page through results
        done = False
        while not done:
            binary = await self.app_client.data_client.binary_data_by_filter(**binary_args)
            if len(binary[0]):
                for b in binary[0]:

                    if b.metadata.file_name != "":
                        file = b.metadata.file_name
                    else:
                        file = b.metadata.id + mimetypes.guess_extension(b.metadata.capture_metadata.mime_type)
                    file = os.path.join(self.mirror_path, file)
                    # check if file exists in target, if so assume its a match and do not re-write
                    if os.path.isfile(file):
                        LOGGER.debug(f"{file} exists already, skipping")
                    else:
                        # get actual binary data
                        id = BinaryID(
                            file_id=b.metadata.id,
                            organization_id=b.metadata.capture_metadata.organization_id,
                            location_id=b.metadata.capture_metadata.location_id
                        )
                        data = await self.app_client.data_client.binary_data_by_ids([id])
                        self.write_file(file, data[0].binary)
                    current_files[file] = False
                # this is where the next page of data will start
                binary_args['last'] = binary[2]
            else:
                done = True
        
        if self.delete:
            # loop through and delete any extra files in the target
            for file in current_files:
                if current_files[file]:
                    os.remove(file)
                    LOGGER.info(f"Deleted {file}")
    
    async def viam_connect(self) -> ViamClient:
        dial_options = DialOptions.with_api_key( 
            api_key=self.api_key,
            api_key_id=self.api_key_id
        )
        return await ViamClient.create_from_dial_options(dial_options)
    
    async def do_command(
            self,
            command: Mapping[str, ValueTypes],
            *,
            timeout: Optional[float] = None,
            **kwargs
        ) -> Mapping[str, ValueTypes]:
        result = {}


    def write_file(self, file_path, content):
        path = Path(file_path)
        
        try:
            # Create the directory and any missing parent directories
            path.parent.mkdir(parents=True, exist_ok=True)
            
            with path.open('wb') as file:
                file.write(content)
            
            LOGGER.info(f"File successfully written: {file_path}")
        except IOError as e:
            LOGGER.error(f"An error occurred while writing the file: {e}")
        except Exception as e:
            LOGGER.error(f"An unexpected error occurred: {e}")

File: mirror/mirror.go
// Package mirror implements the viam-soleng:data:mirror generic service, which
// periodically syncs binary data from Viam's data management down to local files.
package mirror

import (
	"context"
	"errors"
	"fmt"
	"mime"
	"os"
	"path/filepath"
	"slices"
	"sync"
	"syscall"
	"time"

	"go.viam.com/rdk/app"
	"go.viam.com/rdk/logging"
	"go.viam.com/rdk/resource"
	"go.viam.com/rdk/services/generic"
)

// Model is the full triplet for this module's resource.
var Model = resource.NewModel("viam-soleng", "data", "mirror")

const (
	defaultSyncFrequency = 60 * time.Second
	connectionRetries    = 3

	// Preserve 5% of the filesystem. Do not allow this module to fill the drive.
	minFreeFraction = 0.05
)

func init() {
	resource.RegisterService(generic.API, Model, resource.Registration[resource.Resource, *Config]{
		Constructor: newMirror,
	})
}

// Config describes the JSON configuration for the mirror service.
type Config struct {
	AppAPIKey     string   `json:"app_api_key"`
	AppAPIKeyID   string   `json:"app_api_key_id"`
	SyncFrequency float64  `json:"sync_frequency"`
	Tags          []string `json:"tags"`
	Labels        []string `json:"labels"`
	DatasetID     string   `json:"dataset_id"`
	MirrorPath    string   `json:"mirror_path"`
	Delete        bool     `json:"delete"`
	ProtectedDirs []string `json:"protected_dirs"`
}

// Validate ensures the required attributes are present. It returns no dependencies.
func (cfg *Config) Validate(path string) ([]string, []string, error) {
	if cfg.AppAPIKey == "" {
		return nil, nil, resource.NewConfigValidationFieldRequiredError(path, "app_api_key")
	}
	if cfg.AppAPIKeyID == "" {
		return nil, nil, resource.NewConfigValidationFieldRequiredError(path, "app_api_key_id")
	}
	return nil, nil, nil
}

type mirror struct {
	resource.Named
	resource.AlwaysRebuild

	logger logging.Logger

	apiKey        string
	apiKeyID      string
	tags          []string
	labels        []string
	datasetID     string
	mirrorPath    string
	delete        bool
	protectedDirs []string
	syncFrequency time.Duration

	client *app.ViamClient

	workers sync.WaitGroup
	cancel  context.CancelFunc
}

func newMirror(
	ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger,
) (resource.Resource, error) {
	m := &mirror{
		Named:  conf.ResourceName().AsNamed(),
		logger: logger,
	}
	if err := m.reconfigure(conf); err != nil {
		return nil, err
	}

	cancelCtx, cancel := context.WithCancel(context.Background())
	m.cancel = cancel
	m.workers.Add(1)
	go m.syncLoop(cancelCtx)

	return m, nil
}

// reconfigure applies a parsed config to the receiver. It does not start or stop
// the sync loop; that lifecycle is owned by the constructor and Close.
func (m *mirror) reconfigure(conf resource.Config) error {
	cfg, err := resource.NativeConfig[*Config](conf)
	if err != nil {
		return err
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return fmt.Errorf("could not determine home directory: %w", err)
	}

	mirrorPath := filepath.Join(home, ".viam", "data_mirror")
	if cfg.MirrorPath != "" {
		mirrorPath = filepath.Join(home, ".viam", cfg.MirrorPath)
	}

	syncFrequency := defaultSyncFrequency
	if cfg.SyncFrequency > 0 {
		syncFrequency = time.Duration(cfg.SyncFrequency * float64(time.Second))
	}

	m.apiKey = cfg.AppAPIKey
	m.apiKeyID = cfg.AppAPIKeyID
	m.tags = cfg.Tags
	m.labels = cfg.Labels
	m.datasetID = cfg.DatasetID
	m.mirrorPath = mirrorPath
	m.delete = cfg.Delete
	m.protectedDirs = cfg.ProtectedDirs
	m.syncFrequency = syncFrequency

	return nil
}

// DoCommand is unused but required to satisfy the generic service API.
func (m *mirror) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
	return map[string]interface{}{}, nil
}

// Close stops the sync loop and tears down the connection.
func (m *mirror) Close(ctx context.Context) error {
	if m.cancel != nil {
		m.cancel()
	}
	m.workers.Wait()
	m.closeConnection()
	return nil
}

// syncLoop runs do_sync on the configured interval, reconnecting with exponential
// backoff when an error occurs.
func (m *mirror) syncLoop(ctx context.Context) {
	defer m.workers.Done()
	m.logger.Info("Starting sync loop")

	consecutiveFailures := 0
	for {
		if ctx.Err() != nil {
			m.logger.Info("Exiting sync loop")
			return
		}

		if err := m.ensureConnection(ctx); err != nil {
			m.logger.Errorf("Failed to establish connection: %s. Retrying...", err)
			if !sleep(ctx, 5*time.Second) {
				return
			}
			continue
		}

		if err := m.doSync(ctx); err != nil {
			consecutiveFailures++
			m.logger.Errorf("Error in sync loop (attempt %d): %s", consecutiveFailures, err)

			// Force reconnection on error.
			m.closeConnection()

			// Exponential backoff, capped at 60 seconds.
			delay := time.Duration(1<<(consecutiveFailures-1)) * time.Second
			if delay > 60*time.Second {
				delay = 60 * time.Second
			}
			m.logger.Infof("Retrying in %s...", delay)
			if !sleep(ctx, delay) {
				return
			}
			continue
		}

		consecutiveFailures = 0
		if !sleep(ctx, m.syncFrequency) {
			return
		}
	}
}

// sleep waits for d or until the context is cancelled. It returns false if the
// context was cancelled.
func sleep(ctx context.Context, d time.Duration) bool {
	timer := time.NewTimer(d)
	defer timer.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-timer.C:
		return true
	}
}

// ensureConnection makes sure m.client is connected, retrying on failure.
func (m *mirror) ensureConnection(ctx context.Context) error {
	if m.client != nil {
		return nil
	}

	var lastErr error
	for attempt := 0; attempt < connectionRetries; attempt++ {
		m.logger.Infof("Attempting to connect (attempt %d/%d)...", attempt+1, connectionRetries)
		client, err := app.CreateViamClientWithAPIKey(ctx, app.Options{}, m.apiKey, m.apiKeyID, m.logger)
		if err == nil {
			m.client = client
			m.logger.Info("Connection established.")
			return nil
		}
		lastErr = err
		m.logger.Errorf("Connection attempt %d failed: %s", attempt+1, err)
		if attempt < connectionRetries-1 {
			if !sleep(ctx, time.Second) {
				return ctx.Err()
			}
		}
	}
	return fmt.Errorf("failed to connect after %d attempts: %w", connectionRetries, lastErr)
}

// closeConnection closes and clears the active ViamClient, if any.
func (m *mirror) closeConnection() {
	if m.client != nil {
		if err := m.client.Close(); err != nil {
			m.logger.Errorf("Error closing connection: %s", err)
		} else {
			m.logger.Info("Connection closed.")
		}
		m.client = nil
	}
}

// doSync pages through binary data matching the filter, writes any files missing
// from the mirror directory, and (if delete is enabled) removes extra files.
func (m *mirror) doSync(ctx context.Context) error {
	client := m.client
	if client == nil {
		return errors.New("no active connection")
	}
	data := client.DataClient()

	// Build the set of files currently on disk so we can detect extras to remove.
	// A value of true means "not yet seen in data management" (candidate for deletion).
	currentFiles := map[string]bool{}
	if err := filepath.Walk(m.mirrorPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			if os.IsNotExist(err) {
				return nil
			}
			return err
		}
		if !info.IsDir() {
			currentFiles[path] = true
		}
		return nil
	}); err != nil {
		return fmt.Errorf("error walking mirror path: %w", err)
	}

	filter := &app.Filter{}
	if m.datasetID != "" {
		filter.DatasetID = m.datasetID
	}
	if len(m.tags) > 0 {
		filter.TagsFilter = app.TagsFilter{Tags: m.tags}
	}
	if len(m.labels) > 0 {
		filter.BboxLabels = m.labels
	}

	// spaceExhausted is set when a download would breach the free-space reserve.
	// It stops the cycle early and suppresses the delete pass, since currentFiles
	// is then incomplete and cannot be trusted to identify extras.
	spaceExhausted := false

	last := ""
pagingLoop:
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}

		resp, err := data.BinaryDataByFilter(ctx, false, &app.DataByFilterOptions{
			Filter: filter,
			Last:   last,
		})
		if err != nil {
			return fmt.Errorf("error fetching binary data by filter: %w", err)
		}
		if len(resp.BinaryData) == 0 {
			break
		}

		for _, b := range resp.BinaryData {
			if b.Metadata == nil {
				m.logger.Warn("binary data with nil metadata, skipping")
				continue
			}
			file := filepath.Join(m.mirrorPath, fileNameFor(b.Metadata))

			// If the file already exists, assume it is a match and do not re-write it.
			if _, statErr := os.Stat(file); statErr == nil {
				m.logger.Debugf("%s exists already, skipping", file)
				currentFiles[file] = false
				continue
			}

			id := b.Metadata.BinaryDataID
			if id == "" {
				id = b.Metadata.ID
			}
			withData, err := data.BinaryDataByIDs(ctx, []string{id})
			if err != nil {
				return fmt.Errorf("error fetching binary data for %s: %w", id, err)
			}
			if len(withData) == 0 {
				m.logger.Warnf("no binary data returned for %s, skipping", id)
				continue
			}

			content := withData[0].Binary
			room, err := m.hasRoomFor(int64(len(content)))
			if err != nil {
				return fmt.Errorf("failed to check free space on %s before writing %s: %w", m.mirrorPath, file, err)
			}
			if !room {
				if m.fileCanNeverFit(int64(len(content))) {
					m.logger.Errorf("%s (%d bytes) can never fit while preserving %.0f%% free space; skipping",
						file, len(content), minFreeFraction*100)
					continue
				}
				m.logger.Warnf("Pausing sync: writing %s (%d bytes) would drop free space below %.0f%%",
					file, len(content), minFreeFraction*100)
				spaceExhausted = true
				break pagingLoop
			}

			m.writeFile(file, content)
			currentFiles[file] = false
		}

		last = resp.Last
	}

	if m.delete && !spaceExhausted {
		for file, extra := range currentFiles {
			if extra {
				if err := os.Remove(file); err != nil {
					m.logger.Errorf("Error deleting %s: %s", file, err)
				} else {
					m.logger.Infof("Deleted %s", file)
				}
			}
		}
		m.removeEmptyDirs()
	}

	return nil
}

// fileNameFor determines the relative path to write a binary datum to. It mirrors
// the Python module: use the file_name when present, otherwise the data ID plus an
// extension guessed from the MIME type.
func fileNameFor(meta *app.BinaryMetadata) string {
	if meta.FileName != "" {
		return meta.FileName
	}
	ext := meta.FileExt
	if ext == "" {
		if exts, err := mime.ExtensionsByType(meta.CaptureMetadata.MimeType); err == nil && len(exts) > 0 {
			ext = exts[0]
		}
	}
	return meta.ID + ext
}

// hasRoomFor reports whether writing nBytes would still leave at least
// minFreeFraction of total capacity free on the filesystem backing mirrorPath.
func (m *mirror) hasRoomFor(nBytes int64) (bool, error) {
	var st syscall.Statfs_t
	if err := syscall.Statfs(m.mirrorPath, &st); err != nil {
		return false, err
	}
	bsize := int64(st.Bsize)
	total := int64(st.Blocks) * bsize
	avail := int64(st.Bavail) * bsize // space available to non-root users
	reserve := int64(float64(total) * minFreeFraction)
	return avail-nBytes >= reserve, nil
}

// fileCanNeverFit reports whether nBytes exceeds the usable capacity of the
// filesystem (total minus the reserve), i.e. it could not be written even on an
// otherwise empty disk. Such a file would otherwise pause every sync cycle.
func (m *mirror) fileCanNeverFit(nBytes int64) bool {
	var st syscall.Statfs_t
	if err := syscall.Statfs(m.mirrorPath, &st); err != nil {
		return false
	}
	total := int64(st.Blocks) * int64(st.Bsize)
	reserve := int64(float64(total) * minFreeFraction)
	return nBytes > total-reserve
}

// writeFile writes content to file_path, creating parent directories as needed.
func (m *mirror) writeFile(filePath string, content []byte) {
	if err := os.MkdirAll(filepath.Dir(filePath), 0o755); err != nil {
		m.logger.Errorf("An error occurred while creating directories for %s: %s", filePath, err)
		return
	}
	if err := os.WriteFile(filePath, content, 0o644); err != nil {
		m.logger.Errorf("An error occurred while writing the file %s: %s", filePath, err)
		return
	}
	m.logger.Infof("File successfully written: %s", filePath)
}

// removeEmptyDirs removes empty directories under mirrorPath from the bottom up,
// skipping the root and any protected directories.
func (m *mirror) removeEmptyDirs() {
	var dirs []string
	if err := filepath.Walk(m.mirrorPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			if os.IsNotExist(err) {
				return nil
			}
			return err
		}
		if info.IsDir() {
			dirs = append(dirs, path)
		}
		return nil
	}); err != nil {
		m.logger.Errorf("Error walking directories for cleanup: %s", err)
		return
	}

	// filepath.Walk visits parents before children in lexical order, so iterating
	// in reverse processes children before parents (bottom-up), letting a parent
	// be removed in the same pass once its last child directory is gone.
	for i := len(dirs) - 1; i >= 0; i-- {
		dirpath := dirs[i]

		// Skip the root mirror directory.
		if dirpath == m.mirrorPath {
			continue
		}

		relativeDir, err := filepath.Rel(m.mirrorPath, dirpath)
		if err != nil {
			m.logger.Errorf("Error computing relative path for %s: %s", dirpath, err)
			continue
		}
		if m.isProtected(relativeDir) {
			m.logger.Debugf("Skipping deletion for protected directory: %s", dirpath)
			continue
		}

		entries, err := os.ReadDir(dirpath)
		if err != nil {
			m.logger.Errorf("Error reading directory %s: %s", dirpath, err)
			continue
		}
		if len(entries) == 0 {
			if err := os.Remove(dirpath); err != nil {
				m.logger.Errorf("Error removing directory %s: %s", dirpath, err)
			} else {
				m.logger.Infof("Removed empty directory: %s", dirpath)
			}
		}
	}
}

func (m *mirror) isProtected(relativeDir string) bool {
	return slices.Contains(m.protectedDirs, relativeDir)
}
